### ISSUE TITLE
Coordinates of the multi-region mask in the segmentation results of the yolov8m-seg model

### ISSUE THREAD

**Issue by SIG777** [ultralytics/ultralytics#5090]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello.

I'm using the yolov8m-seg segmentation model and analyzing the person class. There are cases where a person is recognized (for example, a person's hand), the rectangular frame is correct, and the mask represents several separate areas (this is normal, because something may be contributing to this).

The mask data is contained in results.masks[i].data[0] and is correct. The coordinates of the polygon that bounds the mask are contained in results.masks.xy[i] and can be uploaded to a markup file (labels). The problem is that the coordinates only represent one of the mask areas, not all mask areas of the recognized person.

I have a question: is it normal that the coordinates are incomplete? At the moment, I found the following solution: get a mask from the YOLO segmentation results, search for contours on it (cv2.findContours) and work with the resulting coordinates. I expected to see the full coordinates of the segmentation polygon in the YOLO results.

Thank you for the opportunity to ask a question and thank you for your work. Best wishes.

**Comment by RizwanMunawar** [comment]
@SIG777, you can utilize the provided code to obtain masks for all detected objects and incorporate them into your specific requirements.

```python
from ultralytics import YOLO

model = YOLO('yolov8n-seg.pt')

# Make predictions with the model
results = model('https://ultralytics.com/images/bus.jpg')

# Extract bounding boxes, classes, object names, confidence scores, and mask
boxes = results[0].boxes.xyxy.tolist()
classes = results[0].boxes.cls.tolist()
names = results[0].names
confidences = results[0].boxes.conf.tolist()
masks = results[0].masks

# Iterate through each detected object's box, class, confidence, and mask
for box, cls, conf, mask in zip(boxes, classes, confidences, masks):
    print(mask)
    # Your additional processing here
    # ...
```

This code allows you to obtain masks for objects detected by the model, which can be further utilized as needed.

**Comment by SIG777** [comment]
@RizwanMunawar, you used a different model. And therefore ended up with different segments.

```
model = YOLO('yolov8n-seg.pt')
```

Please try the one I used. 

```
model = YOLO('yolov8m-seg.pt')
```

**Comment by SIG777** [comment]
@RizwanMunawar, I'm really grateful to you. 

Thank you very much.

**Comment by glenn-jocher** [comment]
@SIG777, it's my pleasure! I'm glad I could assist. Don't hesitate to reach out if you have any other questions or concerns. I hope you find continued success in your work with YOLOv8. Thanks!

**Comment by 8410ttt** [comment]
> ，您可以利用提供的代码获取所有检测到的对象的掩码，并将它们合并到您的特定要求中。
> 
> ```python
> from ultralytics import YOLO
> 
> model = YOLO('yolov8n-seg.pt')
> 
> # Make predictions with the model
> results = model('https://ultralytics.com/images/bus.jpg')
> 
> # Extract bounding boxes, classes, object names, confidence scores, and mask
> boxes = results[0].boxes.xyxy.tolist()
> classes = results[0].boxes.cls.tolist()
> names = results[0].names
> confidences = results[0].boxes.conf.tolist()
> masks = results[0].masks
> 
> # Iterate through each detected object's box, class, confidence, and mask
> for box, cls, conf, mask in zip(boxes, classes, confidences, masks):
>     print(mask)
>     # Your additional processing here
>     # ...
> ```
> 
> 此代码允许您获取模型检测到的对象的掩码，可以根据需要进一步利用这些遮罩。

Hello, when I run the above code I get the following result：
data: tensor([[[0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         ...,
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.]]], device='cuda:0')
orig_shape: (360, 640)
shape: torch.Size([1, 384, 640])
xy: [array([[          0,         165],
       [          0,         191],
       [          1,         191],
       [          3,         193],
       [          3,         194],
       [          4,         195],。。。
xyn: [array([[          0,     0.45833],
       [          0,     0.53056],
       [  0.0015625,     0.53056],
       [  0.0046875,     0.53611],
       [  0.0046875,     0.53889],
       [    0.00625,     0.54167],
       [    0.00625,     0.54444],。。。。
The above result appears 6 , which is exactly the same as the number of masks in the picture. If xy is a mask coordinate, what does xyn represent

**Comment by glenn-jocher** [comment]
@8410ttt hello, thank you for your question. In the output you provided, `xy` and `xyn` both represent coordinates of the object masks. The difference between the two is that `xy` gives the raw pixel coordinates in the image, while `xyn` gives the normalized values of these coordinates.

In `xy`, coordinates are represented in the original pixel space of your image. For example, a coordinate (165,191) means that this point is at the 165th pixel in the width (x) and 191st pixel in the height (y) of your image.

In `xyn`, the same coordinates are represented as a fraction of the image width and height. These values range from 0 to 1. The normalization is done by dividing the x-coordinate by the image width and the y-coordinate by the image height. This representation is useful for making the model generalize better to different image sizes. 

So, in your output, `xyn` is just the normalized version of `xy`. It is the same mask, just represented in a different coordinate space.

**Comment by 8410ttt** [comment]
> ```python
> masks = results[0].masks
> ```
masks = results[0].masks，Can it only be used to extract boundary points? If I need all the pixel coordinates of the mask part, how do I extract them?

**Comment by glenn-jocher** [comment]
@8410ttt yes, the `results[0].masks` syntax is designed to give you the boundary points of the detected objects. The boundary points represent the edges of the object in the image based on the segmentation done by the YOLOv8 model.

To extract all the pixel coordinates of the mask part, you won't be using just the boundary points. Instead, you'll need to use the mask information to generate a binary image, where the mask area is usually shown as white (value of 255), and the non-mask area is shown as black (value of 0). 

Once you have the binary mask image, you can extract the coordinates of all the white pixels. These pixels represent all the pixel coordinates of the segmented object, not just the boundary points. You may use the appropriate image processing tools or libraries in Python to achieve this extraction process.

**Comment by 8410ttt** [comment]
> 是的，语法旨在为您提供检测到的对象的边界点。边界点表示图像中基于YOLOv8模型完成的分割的对象的边缘。`results[0].masks`
> 
> 要提取遮罩部分的所有像素坐标，您不会仅使用边界点。相反，您需要使用遮罩信息来生成二进制图像，其中遮罩区域通常显示为白色（值为 255），非遮罩区域显示为黑色（值为 0）。
> 
> 获得二进制蒙版图像后，您可以提取所有白色像素的坐标。这些像素表示分割对象的所有像素坐标，而不仅仅是边界点。您可以使用 Python 中的相应图像处理工具或库来实现此提取过程。

Okay, thank you very much！！

**Comment by glenn-jocher** [comment]
@8410ttt you're welcome! I'm glad I could help. If you have any further questions or need more clarification, feel free to ask. Happy coding!

**Comment by b-niu** [comment]
In addition, I found that the masks obtained by results[0].masks.data.cpu().numpy() are also wrong. The intersecting detection targets will be truncated. But their bound boxes are correct.

**Comment by b-niu** [comment]
Thank you for your detailed reply, @glenn-jocher . I would be very glad to provide you with some sample data, but unfortunately, these data are medical images, and due to ethical considerations and institutional regulations, I am not authorized to share this data. 😢 I will try to follow your suggestions and do some exploration on the code side. Thank you again.

**Comment by glenn-jocher** [comment]
@b-niu, I completely understand the sensitive nature of medical images and the importance of upholding privacy and ethical standards with such data. If sharing data isn't an option, here are a few suggestions you can explore on the code side, while keeping your data secure:

1. **Model Tuning:** If you have control over model training, fine-tuning the model with a focus on overlapping instances could help. This might include adjusting loss functions to better penalize incorrect mask predictions in overlapping areas.

2. **Data Augmentation:** Including more examples of overlapping instances in the training set, with robust annotations, can help the model learn to handle these scenarios better.

3. **Post-Processing Heuristics:** Develop post-processing scripts that handle overlapping mask segments. For example, you could potentially extrapolate masks within bounding boxes and use heuristics based on the shape and size of known objects to correct them.

4. **Computer Vision Techniques:** Implement advanced computer vision algorithms to process the raw mask outputs from the model, perhaps refining them by considering continuity of object edges or other features characteristic of your specific images.

5. **Secondary Models:** In some cases, a secondary model can assist in refining the outputs of the primary segmentation model. For example, this could be another neural network trained specifically to address mask overlaps.

6. **Review Literature:** Look into recent academic papers and research for novel solutions to instance segmentation with overlapping objects. There might be new methodologies or algorithms you could adapt to your use case.

If there's anything more specific about YOLOv8 or instance segmentation that you'd like to discuss, please let us know. Even without the data, any insights on behaviour or performance patterns can be informative. Your work is contributing to the broader learning of the community, and we appreciate your efforts in exploring solutions within these constraints.

Good luck with your exploration, and don't hesitate to reach out if you have further questions.

**Comment by fdujay** [comment]
@glenn-jocher 
For Data Augmentation, do you think `mixup `or `copy_paste `worked for this problem?

**Comment by b-niu** [comment]
Dear @glenn-jocher , please forgive my rudeness. I would like to ask you a question that may be very simple for you. In the implementation of YOLOV7, if an object's detection frame is offset, the mask predicted for instance segmentation will be cut off by the detection frame, resulting in edges similar to the horizontal and vertical edges shown in the following figure. Will this phenomenon also exist in Yolov8?
![image](https://github.com/ultralytics/ultralytics/assets/32410912/4c746dad-7731-4438-8a64-7f1442681a42)

Has Yolov8 designed a strategy to avoid sharp mask edges that are significantly distorted?
![image](https://github.com/ultralytics/ultralytics/assets/32410912/1a14d062-085b-4e10-a47d-f6c57602e25f)

**Comment by glenn-jocher** [comment]
No need to apologize, @b-niu. Your questions are welcome, and it's great that you're seeking to understand the details of these models.

Firstly, the phenomenon you're describing, where object masks are truncated by the bounding box edges, can occur in any instance segmentation model, including YOLOv7 and YOLOv8, if the bounding boxes predicted by the detection part of the model don't accurately encompass the full extent of the objects.

In YOLOv8, as with many deep learning models, the instance masks are typically generated within the confines of these bounding boxes. Therefore, if the bounding box is misaligned or too small, the mask will be similarly affected, resulting in those sharp, truncated edges.

In terms of strategies to avoid such mask distortion, while there might not be a specific feature within YOLOv8 to address this issue head-on, the precision of mask generation can often be improved by ensuring that the model is well-trained with a sufficiently representative dataset that includes accurate bounding box annotations. Furthermore, advancements in model architecture and training techniques occur over time to improve the quality of both bounding boxes and masks.

It's also worth mentioning that post-processing techniques can be applied to refine the masks. For instance, the raw masks can be expanded to the edges of the respective bounding boxes and then go through a morphological operation (e.g., `dilate` followed by `erode`) to smooth edges and remove the sharp cuts. Moreover, conditional random fields (CRFs) or other graphical models can be used as a post-processing step to refine the segmentation boundaries further.

Please note that while YOLOv8 might share some similarities with previous versions, like the YOLOv7 you mentioned, it also includes updates and improvements, given the active development of deep learning models. So, the manifestation of issues like bounding box and mask alignment might differ between versions, and recent models could potentially handle such cases better due to these continual refinements. However, the underlying challenge of accurately capturing object extents in the presence of occlusion and complex shapes remains an area of ongoing research in the field of computer vision.

**Comment by b-niu** [comment]
Dear @glenn-jocher ,  I have located the problem of missing half of the segment coordinates. On the surface, the problem lies in the `masks2segments(masks, strategy='largest')` function in the `ultralytics/ultralytics/utils/ops.py` file. This function uses the parameter `strategy='largest'` by default, which selects the largest area in the predicted mask and discards the rest. Therefore, when a mask is truncated, part of its segmentation coordinates will be lost.

However, this is only a superficial issue, and the root cause lies in the mask. I used a YOLOV8 model trained on 10,000+ images, but when it predicts masks for objects that overlap, it only assigns the overlapping part to one object, resulting in the other object being missing a piece.

I tried changing the `masks2segments(masks, strategy='largest')` to `masks2segments(masks, strategy='concat')`, but the result was not good, with many incorrect connections being made. It seems that the problem still needs to be addressed at the mask generation step.

Here are some examples that demonstrate the effect of the `concat` parameter.

![image](https://github.com/ultralytics/ultralytics/assets/32410912/12a703f7-7534-488f-9aa0-9670a38db16b)
![image](https://github.com/ultralytics/ultralytics/assets/32410912/37670e9a-4158-4617-a3a1-4e601369424c)
![image](https://github.com/ultralytics/ultralytics/assets/32410912/c7fe0702-95a6-42a8-9338-f3768c14939b)

And the masks are:
![image](https://github.com/ultralytics/ultralytics/assets/32410912/5131062e-0ae7-4533-9d34-f86700e4679b)
![image](https://github.com/ultralytics/ultralytics/assets/32410912/f9640dd5-f43b-46b4-b59c-b9b7393f3b59)
![image](https://github.com/ultralytics/ultralytics/assets/32410912/5465a06a-440d-434d-a1cf-626bc93984a9)

**Comment by glenn-jocher** [comment]
Thank you for sharing your findings, @b-niu. You've identified an important aspect of mask conversion and the limitations of the segmentation algorithm when handling overlapping instances.

When dealing with overlapping objects, assigning the shared region to only one object can lead to missing parts in the masks of other objects. This is an inherent limitation in many segmentation models due to the way they are trained to predict masks. In cases of overlap, the model must decide which object the shared pixels belong to, which may result in the truncation you've observed.

As you noted, switching the `strategy` parameter to `'concat'` in `masks2segments` is not a perfect solution because it can lead to segments being improperly joined. This suggests that a smarter segmentation merge strategy might be needed, which could handle overlaps more intelligently.

Improving the segmentation in areas of overlap may require:

1. **Enhancing the Model's Training Data**: Ensuring that there is enough representation of overlapping objects in the training data with very accurate mask annotations. This helps the model learn to assign pixels correctly in areas of overlap.

2. **Post-Processing Refinement**: After segment conversion, apply advanced computer vision techniques to clean up and refine the segments. This might include operations to disconnect erroneously joined areas, based on knowledge of the expected shapes and sizes of the objects or other domain-specific constraints.

3. **Model Architecture or Loss Function Adjustments**: Potentially modifying the structure of the network or its loss function to better handle the intricacies of overlapping objects, especially for specific applications like medical image analysis where precise segmentations are crucial.

Please keep in mind that instance segmentation remains a challenging research area, particularly in complex scenes with overlapping and closely interacting objects. It's important to continually review the latest literature and consider incorporating new methods that might address these specific challenges.

Although YOLOv8 incorporates many improvements over previous models, it is not immune to the difficulties posed by overlapping objects. The suggestions you're exploring and the issues you've raised are valuable contributions for further developing the capabilities of segmentation models like YOLOv8. 

Keep iterating and refining the techniques used. If you have insights from your continued work on this problem, sharing them, even at a high level, can be beneficial for the community tackling similar challenges.

Your effort in diving deep into these issues is appreciated, and we encourage you to keep pushing forward with the experimentations and optimizations.
